controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackageDiff.java

```diff
@@ -0,0 +1,114 @@
+                .flatMap(contents -> contents.keySet().stream())
+                .sorted()
+                .distinct()
+                .collect(Collectors.toList());
+        for (String file : files) {
+            if (sb.length() > maxTotalDiffSize)
+                sb.append("--- ").append(file).append('\n').append("Diff skipped: Total diff size >").append(maxTotalDiffSize).append("B)\n\n");
+            else
+                diff(Optional.ofNullable(leftContents.get(file)), Optional.ofNullable(rightContents.get(file)), maxDiffSizePerFile)
+                        .ifPresent(diff -> sb.append("--- ").append(file).append('\n').append(diff).append('\n'));
+        }
+
+        return sb.length() == 0 ? "No diff\n" : sb.toString();
+    }
+
+    private static Optional<String> diff(Optional<ZipEntryWithContent> left, Optional<ZipEntryWithContent> right, int maxDiffSizePerFile) {
+        Optional<byte[]> leftContent = left.flatMap(ZipEntryWithContent::content);
+        Optional<byte[]> rightContent = right.flatMap(ZipEntryWithContent::content);
+        if (leftContent.isPresent() && rightContent.isPresent() && Arrays.equals(leftContent.get(), rightContent.get()))
+            return Optional.empty();
+
+        if (left.map(entry -> entry.content().isEmpty()).orElse(false) || right.map(entry -> entry.content().isEmpty()).orElse(false))
+            return Optional.of(String.format("Diff skipped: File too large (%s -> %s)\n",
+                    left.map(e -> e.size() + "B").orElse("new file"), right.map(e -> e.size() + "B").orElse("file deleted")));
+
+        if (leftContent.map(c -> isBinary(c)).or(() -> rightContent.map(c -> isBinary(c))).orElse(false))
+            return Optional.of(String.format("Diff skipped: File is binary (%s -> %s)\n",
+                    left.map(e -> e.size() + "B").orElse("new file"), right.map(e -> e.size() + "B").orElse("file deleted")));
+
+        return LinesComparator.diff(
+                leftContent.map(c -> lines(c)).orElseGet(List::of),
+                rightContent.map(c -> lines(c)).orElseGet(List::of))
+                .map(diff -> diff.length() > maxDiffSizePerFile ? "Diff skipped: Diff too large (" + diff.length() + "B)\n" : diff);
+    }
+
+    private static Map<String, ZipEntryWithContent> readContents(ApplicationPackage app, int maxFileSizeToDiff) {
+        return new ZipStreamReader(new ByteArrayInputStream(app.zippedContent()), entry -> true, maxFileSizeToDiff, false).entries().stream()
+                .collect(Collectors.toMap(entry -> entry.zipEntry().getName(), e -> e));
+    }
+
+    private static List<String> lines(byte[] data) {
+        List<String> lines = new ArrayList<>(Math.min(16, data.length / 100));
+        try (ByteArrayInputStream stream = new ByteArrayInputStream(data);
+             InputStreamReader streamReader = new InputStreamReader(stream, StandardCharsets.UTF_8);
+             BufferedReader bufferedReader = new BufferedReader(streamReader)) {
```

Comment:
It's fine to collapse this to a single declaration, as close will propagate.

Reply:
Simplified.